pkg/resources/cloudcontroller/vsphere.go

```diff
@@ -0,0 +1,162 @@
+
+import (
+	"fmt"
+
+	"k8c.io/kubermatic/v2/pkg/resources"
+	"k8c.io/kubermatic/v2/pkg/resources/reconciling"
+	"k8c.io/kubermatic/v2/pkg/resources/vpnsidecar"
+	"k8c.io/kubermatic/v2/pkg/semver"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+)
+
+const (
+	vsphereCPIControllerManager = "vsphere-cloud-controller-manager"
+)
+
+var (
+	vsphereCPIResourceRequirements = corev1.ResourceRequirements{
+		Requests: corev1.ResourceList{
+			corev1.ResourceMemory: resource.MustParse("100Mi"),
+			corev1.ResourceCPU:    resource.MustParse("200m"),
+		},
+		Limits: corev1.ResourceList{
+			corev1.ResourceMemory: resource.MustParse("512Mi"),
+			corev1.ResourceCPU:    resource.MustParse("500m"),
+		},
+	}
+)
+
+func vsphereDeploymentCreator(data *resources.TemplateData) reconciling.NamedDeploymentCreatorGetter {
+	return func() (string, reconciling.DeploymentCreator) {
+		return vsphereCPIControllerManager, func(dep *appsv1.Deployment) (*appsv1.Deployment, error) {
+			dep.Name = vsphereCPIControllerManager
+			dep.Labels = resources.BaseAppLabels(vsphereCPIControllerManager, nil)
+
+			dep.Spec.Selector = &metav1.LabelSelector{
+				MatchLabels: resources.BaseAppLabels(vsphereCPIControllerManager, nil),
+			}
+
+			podLabels, err := data.GetPodTemplateLabels(vsphereCPIControllerManager, dep.Spec.Template.Spec.Volumes, map[string]string{
+				"component": "cloud-controller-manager",
+				"tier":      "control-plane",
+			})
+			if err != nil {
+				return nil, err
+			}
+			dep.Spec.Template.ObjectMeta = metav1.ObjectMeta{
+				Labels: podLabels,
+			}
+
+			dep.Spec.Template.Spec.DNSPolicy, dep.Spec.Template.Spec.DNSConfig, err =
+				resources.UserClusterDNSPolicyAndConfig(data)
+			if err != nil {
+				return nil, err
+			}
+			dep.Spec.Template.Spec.AutomountServiceAccountToken = pointer.BoolPtr(false)
+			dep.Spec.Template.Spec.HostNetwork = true
+
+			version, err := getVsphereCPIVersion(data.Cluster().Spec.Version)
+			if err != nil {
+				return nil, err
+			}
+			openvpnSidecar, err := vpnsidecar.OpenVPNSidecarContainer(data, openvpnClientContainerName)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get openvpn sidecar: %v", err)
+			}
+			containers, err := getCPIContainerList(version, data)
+			if err != nil {
+				return nil, err
+			}
+			dep.Spec.Template.Spec.Containers = append(containers, *openvpnSidecar)
+
+			dep.Spec.Template.Spec.Volumes = append(getVolumes(), corev1.Volume{
+				Name: resources.CloudConfigConfigMapName,
+				VolumeSource: corev1.VolumeSource{
+					ConfigMap: &corev1.ConfigMapVolumeSource{
+						LocalObjectReference: corev1.LocalObjectReference{
+							Name: resources.CloudConfigConfigMapName,
+						},
+					},
+				},
+			})
+
+			return dep, nil
+		}
+	}
+}
+
+func getCPIContainerList(version string, data *resources.TemplateData) ([]corev1.Container, error) {
```

Comment:
Nit: I don't see that this function can return an error, so I don't think we need to return it.

Reply:
Thanks, I refactored this function a bunch of times and probably I forgot to remove it :smile: